### DISCUSSION TITLE
Google Photos album import with Read-only Gallery

### DISCUSSION THREAD

**Discussion by benooye** [simulot/immich-go#16] [Q&A]
Thanks a lot Jean-François for this amazing project! Indeed this is the missing part of Immich!

A small question to know if immich-go would be compatible with this scenario:
I have already imported my NAS photo in immich with Read-only Gallery feature in immich (with high quality photos), so by not touching the directroy structure and not importing "really" the files in immich (they stay in the NAS folder)
I would like to recreate Google Photos Albums from my Google Takeout archive (medium/less quality photos)
If I use immich-go :
- do the application will properly map the Google Photos Album with the high quality photos already in place in immich?
- only missing photos from my NAS versus Google Takeout will be imported in immich?

Thanks a lot in advance!

Benoît

**Comment by simulot** [comment]
> Thanks a lot Jean-François for this amazing project! Indeed this is the missing part of Immich!

Thank you, you're welcome. I have found the procedure to run immich-cli on my PC not convenient for common people. For some reason I have preferred to write a program instead of installing node.js.

I haven't try your scenario yet. I have to verify how I manage this case. Give me some time to verify that albums can be created after takeout archive bur referencing best quality asset.

Missing photos from immich will be uploaded (copied in immich storage) on the server, and the album created / updated. I have tested this many times.

Also, the feature that replace low quality image by a better one could not working if the  low quality image has been imported in R/O gallery. I need to check this.

I'll give it a try this week-end.
Thanks for testing the program.

JF
Team chocolatine.

**Comment by benooye** [comment]
Thanks for your feedback
I can help to perform tests, my immich installation is not in production yet (I'm under Synology Photos now and would like to switch) so I can offer to perform any tests that could help!

Team chocolatine power too! (Toulouse)

**Reply by simulot** [reply]
Good, you can change the log level to get more information.

**Reply by benooye** [reply]
Nice to see finally Libraries are coming in Immch 1.79 so I will first switch from Read-only Gallery to Librairies, 
then I will try immich-go with the Google Takeout and check the logs

**Reply by simulot** [reply]
I think we can use french for interim messages. 
Je vais aussi tester cette nouvelle version ce weekend si possible.

**Reply by benooye** [reply]
Ok super, je vais également faire des essais ce week-end, voir ce que produisent les logs

**Reply by benooye** [reply]
Voici le résultat d'une première session d'essai avec immich-go en v0.2.4
1/ Sur un serveur immich clean (reset complet) : réinstallé en 1.79.1, chargement des photos existantes (stockage NAS avec photos Hi-Res) via le nouveau mécanisme de Library (et non plus Read-only gallery) dans immich

2/ Lancement de l'import des Google Takeaout avec immich-go commande "upload" avec -log-level=OK
     -> Résultat : 14702 media scanned, 8576 uploaded
     Je n'ai pas trop d'infos sur ce qui s'est passé sur les autres média qui n'ont pas été uplodadé : par exemple est-ce que ce sont tous des existing media / doublons, ou y a t-il des erreurs? Peut-être rajouter cette synthèse serait bien?
    Lors de l'import j'ai eu plusieurs messages du type : 
_can't find the image with title "20170814_193101_HDR.jpg", pattern: "20170814_193101_HDR*.*": file does not exist: "Takeout/Google\u00a0Photos/Saint-Palais été 2017/20170814_193101_HDR                                                        .jpg"_
Je ne sais pas  ce que signifie ce message : est-ce un problème? juste un warning?

3/ Tentative de lancement de nettoyage:
la commande _./immich-go -server=http://xxx:2283 -key=xxx -log-level=OK duplicate -dry-run=yes_
ou la commande _./immich-go -server=http://xxx:2283 -key=xxx duplicate -dry-run=yes_
renvoient tous les 2 les message suivant:

immich-go  0.2.4, commit 17379634937b5e4cee634f410dee207d5b5aab61, built at 2023-09-19T20:06:19Z
Server status: OK
flag provided but not defined: -dry-run
Usage of duplicate:
  -date value
        Process only document having a  capture date in that range. (default 1850-01-04,2030-01-01)
  -yes
        When true, assume Yes to all actions

-> Le process de nettoyage ne se lance pas.

J'ai également essayé de lancer avec la commande _./immich-go -server=http://xxx:2283 -key=xxx -log-level=OK duplicate_
Et là j'ai ce retour:
immich-go  0.2.4, commit 17379634937b5e4cee634f410dee207d5b5aab61, built at 2023-09-19T20:06:19Z
Server status: OK
Get server's assets...GetAllAssets, GET, http://xxxx:2283/api/asset, 200 OK
parsing time "+160895-10-21T05:48:00.000Z" as "2006-01-02T15:04:05Z07:00": cannot parse "+160895-10-21T05:48:00.000Z" as "2006"
parsing time "+160895-10-21T05:48:00.000Z" as "2006-01-02T15:04:05Z07:00": cannot parse "+160895-10-21T05:48:00.000Z" as "2006"

Puis un retour à la ligne de commande de immich-go sans plus d'infos

**Reply by simulot** [reply]
J'ai corrigé aussi les erreurs "time parsing" qui sont causée par des fichiers avec des dates improbables. Immich les enregistrent comme ça, et cela provoque des erreurs à la restitution des pages par le server, et ça faisait planter immich-go.

Maintenant ces erreurs sont ignorées... peut pas faire mieux

**Comment by simulot** [comment]
> Je n'ai pas trop d'infos sur ce qui s'est passé sur les autres média qui n'ont pas été uplodadé : par exemple est-ce que ce sont tous des existing media / doublons, ou y a t-il des erreurs? Peut-être rajouter cette synthèse serait bien?

Lire mon commentaire https://github.com/simulot/immich-go/issues/18#issuecomment-1732258447
Mais oui, pourquoi pas compter les doublons.


>Server status: OK
flag provided but not defined: -dry-run

J'ai changé l'option -dry-run. sans rien préciser c'est un dry run. avec -yes, les suppressions sont tentées


> parsing time "+160895-10-21T05:48:00.000Z" as "2006-01-02T15:04:05Z07:00": cannot parse "+160895-10-21T05:48:00.000Z" as "2006"

Le programme essaye de déterminer la date de prise dans le nom du fichier.
Si le nom de fichier n'est pas une date, le fichier est ouvert pour trover la date dans les données exif ou les metadata des fichiers apple (heic, ou les mov).
Postez l'un des  fichiers concernés si possible.

**Reply by benooye** [reply]
Bonjour, je peux partager le fichier mais je le message d'erreur n'indique pas le nom de celui-ci et je ne sais pas comment le trouver, même avec -log-level=INFO.
Merci d'avance

**Reply by simulot** [reply]
je suis en train de restructurer le code pour contourner les bizarreries des archives takeout.

**Reply by benooye** [reply]
Ok super, de mon côté j'ai installé Go pour pouvoir regarder avec le debugger
Je n'y connais rien en Go, mais on va quand même essayer...

**Reply by benooye** [reply]
Finalement, j'ai pu trouvé le coupable : une vidéo mp4 venant de Google Photos avec un codage bizarre la date correcte est 24 nov. 2016 10:42 UTC+01:00, je ne sais pas comment cela a pu se transformer en "+160895-10-21T05:48:00.000Z"
J'ai supprimé le fichier Google Photo car j'avais l'original qui lui a bien une bonne date...

Bref, je continue avec la commande duplicate et là je tombe sur un nouveau dilemme: il semble que sur certains fichiers (voir tous?), la résolution (0x0), la taille (0B) et la date de prise de vue ( 0001-01-01T00:00:00Z) retournée soit nulle... que ce soit avec des .jpg ou des .mp4:

There are 2 copies of the asset 20150221_171105, taken on 0001-01-01T00:00:00Z
  delete 20150221_171105 0x0, 0 B, /mnt/media/photoBen/Moments/Mobile/LG-H850/DCIM/2016-08-23/20150221_171105.jpg
  keep   20150221_171105 0x0, 0 B, upload/upload/a3762d6f-303e-4917-974a-0ef6b913e29e/896ec9c6-db3d-468c-9bb5-381e4f332a35.jpg

There are 2 copies of the asset 20150222_155900, taken on 0001-01-01T00:00:00Z
  delete 20150222_155900 0x0, 0 B, /mnt/media/photoBen/Moments/Mobile/LG-H850/DCIM/2016-08-23/20150222_155900.mp4
  keep   20150222_155900 0x0, 0 B, upload/upload/a3762d6f-303e-4917-974a-0ef6b913e29e/c5e5fb00-864c-4f97-8999-2363670d6d3a.mp4

Pour rappel, je suis en version Immich v1.79.1
Est-ce un problème déjà connu/identifié?

**Reply by simulot** [reply]
J'ai remarqué le même problème avec une partie de vidéos... je pense que le problème est dans immich lui même...

Je regarderai le second problème dans la semaine.

Merci pour ton aide

**Reply by benooye** [reply]
Effectivement, je viens de regarder en cherchant l'image dans Immich 20150221_171105.jpg
En cliquant sur le (i), j'ai le message suivant:
   Metadata not loaded for /mnt/media/photoBen/Moments/Mobile/LG-H850/DCIM/2016-08-23/20150221_171105.jpg
et sur le fichier en doublons (import Google Photo) j'ai ce message : NO EXIF INFO AVAILABLE
C'est le même problème pour le fichier vidéo 20150222_155900.mp4

Il y a peut être un check à faire dans ce cas et basculer sur une stratégie de comparaison un peu différente? Faire un check sur le taille des fichiers directement à partir du path par exemple?
Et surtout il faudrait avoir la possibilité de choisir le fichier à supprimer, il semblerait que ce soit le premier qui soit proposé à la suppression, hors dans mon cas, je souhaiterai plutôt supprimer le fichier upload de GooglePhotos et conserver le fichier original qui est dans le dossier /mnt/media

**Reply by benooye** [reply]
Version 1.18 d'Immich semble corriger des problèmes d'extraction de metadata
_fix(server): metadata extraction error by @alextran1502 in https://github.com/immich-app/immich/pull/4210_
Je vais voir si le problème est toujours présent

**Reply by benooye** [reply]
Serveur en v1.81.1 opérationnel et clean avec chargement des photos NAS en Library
Est-ce que je lance un import Google Photo Takeout avec la v0.2.4 d'immich-go ou est-ce que ce serait plus utile de tester une future version?
Merci d'avance

**Reply by simulot** [reply]
J'ai migré mon serveur de "prod" hier soir en 1.81.1, et relancé un job extract des métadonnées, mais c'est sans effet.
Je vais tenter d'uploader ces videos avec un fichier sidecar (.xmp) avec les données correctes.

**Reply by benooye** [reply]
Test avec serveur Immich en 1.81.1 et Immich-go v0.2.4
Lancement de l'Import des images Google Takeout pas vu de problème.
Lancement du duplicate, sur les premiers duplicate je n'ai plus le problème d'image avec résolution nulle, taille nulle et date nulle. immich-go rapporte bien les bonnes infos, cf. ci-dessous. Du coup la proposition de suppression est bien la bonne. 
Donc il y avait bien un problème côté Immich et pas Immich-go

immich-go  0.2.4, commit 17379634937b5e4cee634f410dee207d5b5aab61, built at 2023-09-19T20:06:19Z
Server status: OK
Connected, user: xxxxxx
Get server's assets...75728 received
75728 medias, 12232 duplicate(s)...75728 medias, 12232 duplicate(s). Analyze completed.
There are 2 copies of the asset IMG_1542, taken on 2004-04-18T17:49:43Z
  delete IMG_1542 1600x1200, 441.2 KB, upload/library/admin/2004/2004-04-18/IMG_1542.jpg
  keep   IMG_1542 1600x1200, 656.8 KB, /mnt/media/photoBen/Mes Images/2004/2004_04_xxxxxx/04_xxxxx_visite/IMG_1542.jpg
Proceed? [n]/y: y
  Asset removed
There are 2 copies of the asset IMG_1566, taken on 2004-04-18T20:18:25Z
  delete IMG_1566 1600x1200, 366.7 KB, upload/library/admin/2004/2004-04-18/IMG_1566.jpg
  keep   IMG_1566 1600x1200, 581.8 KB, /mnt/media/photoBen/Mes Images/2004/2004_04_xxxxxx/04_xxxxxx_visite/IMG_1566.jpg
...

**Reply by benooye** [reply]
Je viens de forcer le duplicate avec la commande -yes
J'ai vérifié les logs en sortie, sur les 12232 duplicate, plus aucun n'a comme info 0x0 ni 0 B ni 0001-01-01T00:00:00Z
Donc il semble bien que la version Immich 1.81.1 corrige le problème d'image avec résolution nulle, taille nulle et date nulle

**Reply by simulot** [reply]
Salut, 
Après 2 semaines de déplacement, et divers éventements familiaux, j'ai bossé sur le programme ce week end. 
J'ai réécrit la gestion du takeout. La version est mergée dans la branche dev

Je ferai une release dans quelques jours si je ne trouve pas de problème

**Reply by benooye** [reply]
Ok super, prêt à tester!
Pour info, j'ai constaté grâce à Immigch-go que j'avais pas mal de doublons dans ma Library qui est uniquement en read-only pour Immich.
Du coup immich-go détecte des duplicate et essaie de les supprimer sans succès.
Pas de soucis pour moi c''est juste pour info et d'ailleurs il n'y a pas de message d'erreur. 
C'est juste qu'en relançant immich-go j'ai pu être surpris de voir qu'il y avait encore des doublons mais donc c'est tout à fait normal.

**Reply by simulot** [reply]
A vérifier quand même. Immich server utilise un hash SHA1 pour détecter les doublons. Donc, c'est très fiable pour détecter les copies de fichiers même ayant des noms différents.

Reste les doublons engendrés par des différences de résolutions. C'est ce que immich-go signale.

**Reply by benooye** [reply]
Merci pour le travail que tu fais, le projet est vraiment super et utile pour tous ceux qui veulent migrer.
Pour info, v0.3.1 lancée hier en import puis duplicate je n'ai pas noté de soucis particuliers.
Par contre j'ai perdu les logs, je voulais quand même regarder un peu plus dans le détail.
Je vois de nouveaux albums importés avec des noms de personnes, apparemment avec des photos qui avaient été partagées sous Google Photos

**Reply by simulot** [reply]
Merci!
>Je vois de nouveaux albums importés avec des noms de personnes, apparemment avec des photos qui avaient été partagées sous Google Photos

Donc ça marche pas mal
